pkg/reconciler/pipelinerun/affinity_assistant_test.go

```diff
@@ -44,43 +44,164 @@ import (
-			Spec: v1beta1.PipelineRunSpec{
-				Workspaces: []v1beta1.WorkspaceBinding{{
-					Name: "PersistentVolumeClaim Workspace",
-					PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{
-						ClaimName: "myclaim",
-					},
-				}},
+		pr:   testPRWithPVC,
+		expectStatefulSetSpec: &appsv1.StatefulSetSpec{
+			Template: corev1.PodTemplateSpec{
+				Spec: corev1.PodSpec{
+					Volumes: []corev1.Volume{{
+						Name: "workspace-0",
+						VolumeSource: corev1.VolumeSource{
+							PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{ClaimName: "myclaim"},
+						},
+					}},
+				},
+			},
+		},
+	}, {
+		name: "VolumeClaimTemplate Workspace type",
+		pr:   testPRWithVolumeClaimTemplate,
+		expectStatefulSetSpec: &appsv1.StatefulSetSpec{
+			VolumeClaimTemplates: []corev1.PersistentVolumeClaim{{
+				ObjectMeta: metav1.ObjectMeta{Name: "pvc-b9eea16dce"},
+			}},
+		},
+	}, {
+		name: "VolumeClaimTemplate and PersistentVolumeClaim Workspaces",
+		pr:   testPRWithVolumeClaimTemplateAndPVC,
+		expectStatefulSetSpec: &appsv1.StatefulSetSpec{
+			VolumeClaimTemplates: []corev1.PersistentVolumeClaim{{
+				ObjectMeta: metav1.ObjectMeta{Name: "pvc-b9eea16dce"},
+			}},
+			Template: corev1.PodTemplateSpec{
+				Spec: corev1.PodSpec{
+					Volumes: []corev1.Volume{{
+						Name: "workspace-0",
+						VolumeSource: corev1.VolumeSource{
+							PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{ClaimName: "myclaim"},
+						},
+					}},
+				},
 			},
 		},
+	}, {
+		name:                  "other Workspace type",
+		pr:                    testPRWithEmptyDir,
+		expectStatefulSetSpec: nil,
+	}}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			ctx := context.Background()
+			ctx, cancel := context.WithCancel(ctx)
+			defer cancel()
+
+			c := Reconciler{
+				KubeClientSet: fakek8s.NewSimpleClientset(),
+			}
+
+			err := c.createOrUpdateAffinityAssistantsPerPipelineRun(ctx, tc.pr.Spec.Workspaces, tc.pr, tc.pr.Namespace)
+			if err != nil {
+				t.Errorf("unexpected error from createOrUpdateAffinityAssistantsPerPipelineRun: %v", err)
+			}
+
+			// validate StatefulSets from Affinity Assistant
+			expectAAName := getAffinityAssistantName("", tc.pr.Name)
+			aa, err := c.KubeClientSet.AppsV1().StatefulSets(testPRWithPVC.Namespace).Get(ctx, expectAAName, metav1.GetOptions{})
+			if tc.expectStatefulSetSpec != nil {
+				if err != nil {
+					t.Fatalf("unexpected error when retrieving StatefulSet: %v", err)
+				}
+				if d := cmp.Diff(tc.expectStatefulSetSpec, &aa.Spec, statefulSetSpecFilter, podSpecFilter, podTemplateSpecFilter); d != "" {
+					t.Errorf("StatefulSetSpec diff: %s", diff.PrintWantGot(d))
+				}
+			} else if !apierrors.IsNotFound(err) {
+				t.Errorf("unexpected error when retrieving StatefulSet which expects nil: %v", err)
+			}
+
+			// TODO: test cleanupAffinityAssistants for coscheduling-pipelinerun mode when fully implemented
```

Comment:
is there a way to implement support for cleanup within this PR?

Reply:
I'm planning to implement the cleanup logic for coscheduling-pipelinerun mode in a separate PR to make review easier. I have added an todo to the `cleanupAffinityAssistants ` function